### PR TITLE
feat: detect PVC StatefulSet ownership

### PR DESCRIPTION
**Pull request by alexmt**
PR introduces dynamic discovery of StatefulSet to PVC ownership. Closes https://github.com/argoproj/argo-cd/issues/2924

### CODE REVIEW

File: pkg/cache/references.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+var (
+	pvcNameRegex = regexp.MustCompile(`.+-(.+)-\d+`)
+)
+
+func (c *clusterCache) resolveResourceReferences(un *unstructured.Unstructured) ([]metav1.OwnerReference, func(kube.ResourceKey) bool) {
+	var isChildRef func(_ kube.ResourceKey) bool
+	ownerRefs := un.GetOwnerReferences()
+	gvk := un.GroupVersionKind()
+	// Special case for endpoint. Remove after https://github.com/kubernetes/kubernetes/issues/28483 is fixed
+	if gvk.Group == "" && gvk.Kind == kube.EndpointsKind && len(un.GetOwnerReferences()) == 0 {
```
Comment (ash2k) [inline pkg/cache/references.go:27]: nitpick: all these `if`s can probably be turned into a `switch` for better readability.

Comment (alexmt) [reply pkg/cache/references.go]: Thank you for the suggestion! Switch is actually easier to read. Updated.

**Comment by opensourceonly** [comment]
@alexmt Do we know if this will be available in which version of Argo CD ?

**Comment by alexmt** [comment]
@opensourceonly  We plan to release is as part of Argo CD v1.7

**Comment by opensourceonly** [comment]
@alexmt i hope 1.7 is not far away to be released

**Comment by eric-sellers** [comment]
Hello @alexmt ,

I would appreciate your confirmation regarding the following.

Since this issue did not make it in to gitops-engine v0.1.3, am I correct in my assumption that Argo CD v1.7 does *not* detect PVC StatefulSet ownership? 

As far as I have experimented by deploying applications that have StatefulSets using ArgoCD v1.6 and v1.7, both show the exact same behaviour, where the PVC appears to be unrelated to the StatefulSet and is marked as "Should be pruned".

If my understanding is correct, the claim that this issue is resolved in v1.7 as announced in the following page is incorrect.

- https://github.com/argoproj/argo-cd/releases
  - feat: detect PVC StatefulSet ownership (#112)

**Comment by alexmt** [comment]
Hello @eric-sellers , it is part of 1.7 but has a bug. Just sent PR with https://github.com/argoproj/gitops-engine/pull/138

**Comment by eric-sellers** [comment]
Hello alexmt,

I see. Thank you for updating me on the latest situation.

**Comment by opensourceonly** [comment]
@alexmt is the bug fix part of a release other than 1.7 ?

**Pull request merged**